akka-docs/src/main/paradox/scala/stream/stream-substream.md

```diff
@@ -0,0 +1,125 @@
+
+If you add a `Sink` or `Flow` right after the `groupBy` stage,
+all transformations are applied to all encountered substreams in the same fashion.
+So, if you add the following `Sink`, that is added to each of the substreams as in the below diagram.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy2 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy2 }
+
+![stream-substream-groupBy2.png](../../images/stream-substream-groupBy2.png)
+
+Also substreams, more precisely, `SubFlow` and `SubSource` have methods that allow you to
+merge or concat substreams into the master stream again.
+
+The `mergeSubstreams` method merges an unbounded number of substreams back to the master stream.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy3 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy3 }
+
+![stream-substream-groupBy3.png](../../images/stream-substream-groupBy3.png)
+
+You can limit the number of active substreams running and being merged at a time,
+with either the `mergeSubstreamsWithParallelism` or `concatSubstreams` method.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy4 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy4 }
+
+However, since the number of active substreams is capped,
+be careful so that these methods do not cause deadlocks with back pressure like in the below diagram.
+
+![stream-substream-groupBy4.png](../../images/stream-substream-groupBy4.png)
+
+### splitWhen and splitAfter
+
+`splitWhen` and `splitAfter` are two other operations which generate substreams.
+
+The difference from `groupBy` is that, if the predicate for `splitWhen` and `splitAfter` returns true,
+a new substream is generated, and the succeeding elements after split will flow into the new substream.
+
+`splitWhen` flows the element on which the predicate returned true to a new substream,
+ whereas `splitAfter` flows the next element to the new substream after the element on which predicate returned true.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #splitWhenAfter }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #splitWhenAfter }
+
+![stream-substream-splitWhen-splitAfter.png](../../images/stream-substream-splitWhen-splitAfter.png)
+
+## Flattening stages
+
+### flatMapConcat
+
+`flatMapConcat` and `flatMapMerge` are substream operations different from `groupBy` and `splitWhen/After`.
+
+`flatMapConcat` takes a so-called predicate, which is `f` in the following diagram.
+The predicate `f` of `flatMapConcat` transforms each input element into a `Source` that is then flattened
```

Comment:
That is not a predicate, a predicate is always `A => Boolean`. This is just a function.